### DISCUSSION TITLE
IMU (gyro) drifts over time, consistently

### DISCUSSION THREAD

**Discussion by hawkina** [HumanSupportRobot/QandA#94] [Q&A]
Hey guys,
we seem to have an issue with our base_imu sensor and are not sure on how to fix it. 
**Symptoms:** after standing still for 10+ minutes, one can see in Rviz that the laser scan no longer matches the wall, even though the robot has been previously localized correctly (see screenshot). It seems like the laser has rotated over time. Since the robot has not been moved, AMCL doesn't notice this drift and therefore does not correct the pose. We tried to find the source of the issue and have noticed, that the base_imu seems to be drifting over time, consistently, into one direction.
**Drifting topic**: /hsrb/base_imu/data/ **value:** orientation/z
Tried solutions: we tried to call the re-calibration service: /hsrb/base_imu/calibrate and it forces it back to 0, but then it still keeps on falling. I have attached some screenshots below. 
Rviz drift: 
![hsr_rviz_drift](https://github.com/HumanSupportRobot/QandA/assets/18050791/3fa2e058-2b42-42e2-8910-d7e63c1c9fd2)

plot of /hsrb/base_imu/data/orientation/z:
![hsr_base_imu_orientation_z](https://github.com/HumanSupportRobot/QandA/assets/18050791/96b921fe-82fa-4109-820a-25793bb7517b)
trying to re-calibrate:
![imu_recalibrate_1](https://github.com/HumanSupportRobot/QandA/assets/18050791/6af8e9be-24ac-4d91-b2cc-74cdf1be33d5)

We don't think it is sensor noise since it is really consistent and we didn't have a problem with it before.
I could think of a few hacky solutions to this problem, by remapping ros topics and counteracting this issue but maybe there is a way of fixing this? Is it a reason to get worried, open up the robot and investigating further? I have seen a few roslaunch files on the hsr about imu, imu bias and similar... might be a possible solution also?

We would really appreciate some insights or potential solutions. Thank you in advance!

**Comment by yoshimi-iyoda** [comment]
With the default settings of HSR, the imu topics are not used for anything, so we did not encounter any problems with laser scans and wall misalignment, and we were not too concerned about the imu drift issue.

The drift of the HSR imu sensor I tested was similar to yours, as shown in the figure below.

![image](https://github.com/HumanSupportRobot/QandA/assets/71624812/bd65a244-07de-4694-bcce-188cd7a59189)

A filter might help you avoid the problem. I will try the [Madgwick filter](https://wiki.ros.org/imu_filter_madgwick) in imu_tools to see if it works.

**Reply by yoshimi-iyoda** [reply]
When I check imu's drift again with PlotJugger, it is as follows, and it seems that the yaw deviation is not so large, about 0.1[deg/h].

![image](https://github.com/HumanSupportRobot/QandA/assets/71624812/5b9c47c7-e9cb-4c2b-94ee-4fa7b5c6daa5)

I also checked the axis direction of imu with rviz using imu plugin and found that there is not much misalignment.

Some time has passed since the HSR was started
![image](https://github.com/HumanSupportRobot/QandA/assets/71624812/3d7345d3-9898-4d22-9880-a7911d19ba25)

After 70 minutes
![image](https://github.com/HumanSupportRobot/QandA/assets/71624812/768cc1cb-953f-46d4-af41-25de67b18c04)

So, I would like to ask, how do you use imu data for self-location?
Well, if you don't mind, could you please share with me the software you are using?

**Reply by hawkina** [reply]
Thanks for the replies.
We were also considering adding a filter, but wanted to check if re calibration of the sensor would be possible at all. 
We are using the default ros amcl localization. The tuned parameters we use can be seen here: https://github.com/code-iai/iai_hsr/blob/main/hsr_navigation/launch/hsr_amcl.launch
We opted for amcl since last time when we went with the default one during the competition, the robot got lost within the arena and did not align its position based on the laser scanner again, and we had a large rotational offset. Amcl re-localizes a lot more frequently and seems to be more stable. If the robot keeps driving, then this rotational drift gets corrected by the localization. Even if the offset is accumulated, as soon as it drives, it usually fixes itself - but still, since it could fail, I would like to fix it. 

I think the IMU contributes to the odometry, and since amcl uses it aswell, this is how the error sneaks in. 
I guess I could either look more deeply into amcl parameters to try and fix it, if that doesn't work, a workaround would be an additional filter or maybe a localization trigger over time.

**Reply by yoshimi-iyoda** [reply]
I just asked someone in the know,

http://github.com/code-iai/iai_hsr/blob/main/hsr_navigation/launch/hsr_amcl.launch

Regarding,
The person said that adjusting odom_alpha1 larger (like above 1) and update_min_a smaller might improve it, but it might cause other inconveniences.

I apologize for not being able to help you.

If we notice anything separately, we will respond.
